### ISSUE TITLE
Discussion: Client-side unit testing

### ISSUE THREAD

**Issue by sebastianhoitz** [tower-archive/tower#112]
This is intended as an open discussion about the best way to do client-side unit testing.

We have to think about a way to do unit testing for the client-side.

We need this for quite a few parts. Client-side view rendering, socket.io integration, routing of actions to the controller, to name just a few.

For client-side unit tests there are a lot of different approaches.
- You could use [PhantomJS](http://code.google.com/p/phantomjs/) to do headless, webkit testing and actually open a normal browser to do the tests. The problem with this is, that it does complete infrastructure tests. That means, that even if the error is on the server-side, the client-side might still fail, because we get an unexpected result.
- We could unit test the client-side controllers, models and renderers separately. The problem with this right now is, that the server-side controllers overwrite the client-side controllers, since all of them write to the same namespace (App.*Controller).
- Another suggestion for headless webkit testing: [zombieJS](http://zombie.labnotes.org/)

One problem we also have to be aware of:

For some reason, unit testing socket.io on the client-side proves tricky. My out-of-the-dark guess is, that asynchronously executing the tests lets tower and socket.io block itself on the different ports. But this is just a guess, and not yet proven.

What implementation do you suggest @viatropos ?

**Comment by lancejpollard** [comment]
Great discussion topic.

My preferred way of testing the client is through design.io + mocha.  Say you have your app open at a specific URL like "/login".  You write integration tests testing that the login form has the right labels and highlights the field in red when an input looses focus or whatever.  You can also test that the login form has a specific width, i.e. test all of the css is rendered correctly.

Using web sockets like this, you can write tests for your app **with your page open** and when you hit save it will reload the javascript file your tests are written in, and mocha will re-execute them.  This is very easy to do, we just need to create a `watch` task for design.io (have done this a bunch before, haven't had time to refactor it out).

``` coffeescript
mint = require 'mint'

watch /test\/*Test.coffee/
  # this is in the Watchfile context
  update: (path, callback) ->
    mint.render path: path, (error, result) =>
      @broadcast body: result
      callback()

  client:
    # this is in the browser context
    update: (data) ->
      eval("(#{data.body})") # this is the mocha test code you just saved...

      mocha.run()
```

This way, you can also test multiple-page flows. In an `after` block in your mocha test, you can do `window.location = x`, or in one of your tests you can submit a non-ajax form, and if the page changes, it will load the next url and if there are any tests associated with that url, those will run!  That's pretty powerful.  You could create a demo of your app this way, where it looks like a bot is basically performing every action possible on your app, changing pages and everything.

Another big benefit is you can have multiple different browser windows open (firefox, opera, etc.) and because of design.io the tests will execute in all of the browsers open!

You could probably even do that with Internet Explorer through VMWare fusion or Parallels (I use VMWare fusion now).

I've used PhantomJS and ZombieJS (and selenium) in the past, and while they are nice, it never _really_ mimics the browser environment...  You end up not being able to test fixes for those cross-browser errors, and testing async/visual javascript is not practical in a headless environment.

We should probably have a way to test using PhantomJS and ZombieJS anyway though, because having a browser window open while testing can be annoying sometimes.  But there are other ways around this.

Also, it's very easy to programmatically open browser windows from the terminal and run tests, so the design.io/watch solution definitely works as part of a test suite.

We also need to consider auto-executing tests on file save in the terminal (like [Spork](https://github.com/sporkrb/spork) in Rails).  In practice though, I found spork very hard to setup and it was rather annoying having tests auto-run every time you save a file.  The way I test, I usually run 1 test file for the features I'm working on, and every 10 or so changes I run the suite, so if anything breaks I catch it quickly.  This isn't a matter of the suite taking a long time to run, it's more a matter of not being overwhelmed by output (even 750 dots........................ distracts my mind, I'd rather only see the dots I'm focused on at the moment).

For client-side tests though, I think we should go with the in-browser solution using design.io.  It's SUPER sick.  Everybody's doing headless tests but they don't cover a lot of the test cases you really should be testing from a UI/client javascript perspective.

To do this, we need:
1. a watch task for design.io
2. a web console reporter for mocha (outputs the test results in the web console using `console.group`, very cool.

I implemented this for Jasmine a while ago, here's the code:

http://github.com/viatropos/design.io-jasmine
https://github.com/viatropos/design.io-jasmine/blob/master/src/design.io-jasmine.coffee

Basically, we just create a plugin design.io-mocha that takes care of the details.

What do you think?

**Comment by lancejpollard** [comment]
p.s. I'm a fan of doing discussions like this, let's just label issues with `discussion` and we'll close them once we feel the discussion reaches a reasonable conclusion.  You can always comment on closed discussions or reopen them.  Forget google groups :)

**Comment by edubkendo** [comment]
Two things-- mocha has a watch option to autotest-on-save that's baked in, I actually really like testing this way because I also use the growl option so that I can focus on coding or whatever while tests are running, then get a quick notification through libnotify (but mocha uses growl for mac or windows) that just pops up in the corner of the screen with my results pass/fail. Then, if its a fail I can go look more closely at the output in the terminal to see why. 

The open-browser solution your talking about sounds pretty awesome, but I'd like a headless option as well simply because it kind of depends WHAT your testing on the client side whether you really need a real browser or not. Also, as far as I can tell, design.io still only works for Mac folks right?  

In the js/node world those of us using linux might be the minority, but I think we're a BIG minority.  I don't mind not being able to use design.io too much, I mean, I would REALLY like to be able to use it, but I can live without it. I can't live without being able to easily set up automated testing on client side code. And it sounds like, without design.io, unless we have some sort of headless option baked in, that's pretty much what the situation would be.

**Comment by lancejpollard** [comment]
Glad we got design.io working on linux!  I agree, we should have a headless option as well.

**Comment by sebastianhoitz** [comment]
So, @viatropos do you think you could implement the design.io testing on the client-side? I assume you know best how to implement that. ;)

And then @edubkendo or I could implement headless testing. I think I prefer zombieJS, because it looks like it would better fit the unit-testing style.

PhantomJS looks more like web-scraping. What are your thoughts?

PS: I wasn't allowed to add labels to the issue. Otherwise I would have :)

**Comment by edubkendo** [comment]
For what it's worth:
zombieJS appears to be quite easy to use with Mocha, ie https://github.com/mizchi/zombie-tester-example .  Whereas using PhantomJS with Mocha seems to be a more difficult proposition: https://github.com/visionmedia/mocha/issues/160 .

**Comment by lancejpollard** [comment]
@sebastianhoitz sounds good.  I won't be able to get to it for a couple weeks though, need to get this ember/client stuff finished.

I agree with you on PhantomJS as well, I've only used it to run the ember test suite and it just didn't feel as nice as zombie.js, no idea why.  I like zombie, have enjoyed using it for other things.

Perhaps we should think about building something like [Capybara](https://github.com/jnicklas/capybara) for node - at least an "interface" to other integration testing frameworks/tools.  Not the biggest fan of having a super rich DSL for test driven development, but Capybara is an excellent project.

@edubkendo Mocha is the best. Mocha + Zombie will be great.

P.S.  The main reason I like doing integration testing in the browser (via design.io) instead of headless with zombie is because of the XPath problem.  Testing if certain HTML exists is MUCH easier with jQuery than trying to build another XPath parser with a _different_ API.  In Capybara/Selenium, you had to do this:

``` ruby
page.has_xpath?('//table/tr')
```

Whereas in jQuery I could do this:

``` javascript
$('table tr').length
```

Most people don't even know jQuery has a "css interface" over XPath, to query HTML.  They just traverse the DOM the way they intuitively know how to do.  With Capybara/Selenium, it felt like you had to be a master of XPath before you could even start testing.  We should definitely avoid that, make it easy for everyone.

If you can figure out how to use jQuery in a headless browser, at least _just_ to take advantage of it's XPath/CSS selector system for testing HTML, that would be **awesome**.  I think I've seen something like this but haven't tried.

**Comment by edubkendo** [comment]
zombie actually implements sizzle.js for CSS selectors, which is also what jquery uses. It looks like this comes pretty close to what you are wanting, though I could certainly be misunderstanding. From the [zombiejs homepage](http://zombie.labnotes.org/) :

> There are several ways you can inspect the contents of a document. For starters, there's the DOM API, which you can use to find > elements and traverse the document tree.
> 
> You can also use CSS selectors to pick a specific element or node list. Zombie.js implements the DOM Selector API. 
> These functions are available from every element, the document, and the Browser object itself...

``` javascript
// Make sure we have an element with the ID brains.
assert.ok(browser.query("#brains"));

// Make sure body has two elements with the class hand.
assert.lengthOf(browser.body.queryAll(".hand"), 2);

// Check the document title.
assert.equal(browser.text("title"), "The Living Dead");

// Show me the document contents.
console.log(browser.html());

// Show me the contents of the parts table:
console.log(browser.html("table.parts"));
```

> CSS selectors are implemented by Sizzle.js. In addition to CSS 3 selectors you get additional and quite useful
> extensions, such as :not(selector), [NAME!=VALUE], :contains(TEXT), :first/:last and so forth. Check out the Sizzle.js
> documentation for more details.

So, if I haven't misunderstood, I think you can get what you want from zombie fairly easily, which would be awesome, as I agree that is much easier to use.
